File: src/Graph/EdmondsKarp.java
import java.util.Queue;
import java.util.Arrays;
import java.util.LinkedList;


class EdmondsKarp {
  static boolean augmentingPath(int[][] Gf, int s, int t, int[] parent) {
    // standard BFS starting at s on Gf, storing parent of vertices in parent array

    int V = Gf.length;

    boolean[] visited = new boolean[V];
    Arrays.fill(visited, false);

    Queue<Integer> Q = new LinkedList<>();
    Q.add(s);
    visited[s] = true;

    while (!Q.isEmpty()) {
      int u = Q.poll();

      for (int v = 0; v < V; v++) {
        if (Gf[u][v] > 0 && !visited[v]) {
          Q.add(v);
          parent[v] = u;
          visited[v] = true;
        }
      }
    }

    return visited[t];
  }

  // Computes the max flow in the flow network Gf
  // where Gf is an adjacency-matrix representation of the network
  // Note Gf is mutated by this procedure.
  //
  // Implementation based off of wikipedia: https://en.wikipedia.org/wiki/Ford%E2%80%93Fulkerson_algorithm
  // Time Complexity: O(VE^2)
  static int apply(int[][] Gf, int source, int sink) {
    int V = Gf.length;
    int[] parent = new int[V];

    int maxflow = 0;

    while (augmentingPath(Gf, source, sink, parent)) {
      long pathflow = Integer.MAX_VALUE;
      int s = sink;

      // find augmenting path's bottleneck
      while (s != source) {
        pathflow = Math.min(pathflow, Gf[parent[s]][s]);
        s = parent[s];
      }

      maxflow += pathflow;

      // update residual capacities in Gf along augmenting path
      int v = sink;
      while (v != source) {
        Gf[parent[v]][v] -= pathflow;
        Gf[v][parent[v]] += pathflow;
        v = parent[v];
      }
    }

    return maxflow;
  }
}

File: src/Graph/edmonds_karp.cpp
#include<climits>
#include<algorithm>
#include<vector>
#include<queue>
using namespace std;

bool augmenting_path(vector<vector<int>>& Gf, int s, int t, vector<int>& parent) {
  // standard BFS starting at s on Gf, storing parent vertices in parent vector
  int V = Gf.size();

  vector<bool> visited(V);
  queue<int> Q;
  Q.push(s);
  visited[s] = true;

  while (!Q.empty()) {
    int u = Q.front();
    Q.pop();

    for (int v = 0; v < V; v++) {
      if (Gf[u][v] > 0 && !visited[v]) {
        Q.push(v);
        parent[v] = u;
        visited[v] = true;
      }
    }
  }

  return visited[t];
}

/**
 * Computes the max flow in the flow network Gf
 * where Gf[u][v] is the residual capcity of edge (u, v).
 * Note Gf is mutated by this procedure.
 *
 * Implementation based off of wikipedia.
 * Time Complexity: O(|V|*|E|^2)
 */
int edmonds_karp(vector<vector<int>>& Gf, int source, int sink) {
  int V = Gf.size();
  vector<int> parent(V);

  int maxflow = 0;

  while (augmenting_path(Gf, source, sink, parent)) {
    long long pathflow = INT_MAX;
    int s = sink;

    // find augmenting path's bottleneck
    while (s != source) {
      pathflow = min(pathflow, (long long) Gf[parent[s]][s]);
      s = parent[s];
    }

    maxflow += pathflow;

    // update residual capacities along augmenting path
    int v = sink;
    while (v != source) {
      Gf[parent[v]][v] -= pathflow;
      Gf[v][parent[v]] += pathflow;
      v = parent[v];
    }
  }

  return maxflow;
}
